src/FileInputFieldWithList.js

```diff
@@ -0,0 +1,152 @@
+import React, { Component } from 'react'
+import propTypes from '@dhis2/prop-types'
+
+import { FileInputField } from './FileInputField.js'
+import { FileListItemWithRemove } from './FileInputFieldWithList/FileListItemWithRemove.js'
+import { statusPropType, sizePropType } from './common-prop-types.js'
+
+/* 
+NOTE:
+A JavaScript FileList instance is read-only, so we cannot add files to it
+FileList also doesn't have a .map method so by destructuring the FileList
+instance into an array we can add, remove and map
+*/
+
+class FileInputFieldWithList extends Component {
+    handleChange = ({ files: fileList }, event) => {
+        const { onChange, name } = this.props
+
+        onChange({ files: this.updateFileArray(fileList), name: name }, event)
+    }
+
+    updateFileArray(fileList) {
+        const { multiple, files } = this.props
+
+        if (!multiple) {
+            // Spread immutable FileList instance onto array
+            return [...fileList]
+        }
+
+        return [...files, ...fileList].reduceRight(
+            (acc, file) => {
+                if (!acc.unique.has(file.name)) {
+                    acc.unique.add(file.name)
+                    acc.files.unshift(file)
+                }
+                return acc
+            },
+            { files: [], unique: new Set() }
+        ).files
+    }
+
+    handleRemove = ({ file: fileToRemove }, event) => {
+        const { files, onChange, name } = this.props
+
+        onChange(
+            {
+                files: files.filter(file => file !== fileToRemove),
+                name,
+            },
+            event
+        )
+    }
+
+    render() {
+        const {
+            accept,
+            buttonLabel,
+            className,
+            dataTest,
+            disabled,
+            error,
+            files,
+            helpText,
+            initialFocus,
+            label,
+            large,
+            multiple,
+            name,
+            onBlur,
+            onFocus,
+            placeholder,
+            removeText,
+            required,
+            small,
+            tabIndex,
+            valid,
+            validationText,
+            warning,
+        } = this.props
+
+        return (
+            <FileInputField
+                accept={accept}
+                buttonLabel={buttonLabel}
+                className={className}
+                dataTest={dataTest}
+                disabled={disabled || (!multiple && files.length >= 1)}
```

Comment:
This means that the button is disabled once a file has been selected when multiple is not true.
When thinking about people who are not as versatile with how the web works as we are, then they might think, that they can't replace the file anymore, because the button is disabled, all they can do is removing the existing file.

What I really want to say is: This way seem correct/good from a technical perspective, but it requires the user to think, which is not so good imo.

@cooper-joe You're our UX guy, what do you think about that?

Reply:
I think the current implementation is the simplest solution. `Replacing a file` is a more advanced concept than `Add>Delete>Add again`. 

Allowing file replacement would open a more difficult problem: how do we inform to the user that only one file is allowed, even though they can "upload" many? The simpler solution is to ensure the user understands the concept by enforcing the `Add > Delete > Add again` method, thereby avoiding any confusion with one vs. multiple files being uploaded.

Disabling the button is intentionally slowing down the user, highlighting to them that they can’t upload more files without removing the one that is already there.

Reply:
Fair enough. Now I'm wondering if there's a way now we can let the user know that he has to `Add > Delete > Add`. Maybe by changing the text of the button (optionally by default, app could override this behavior) or by displaying a helper text once a file has been selected like "If you want to add a different file instead, please remove the existing one first"